safir/src/safir/metrics/_event_manager.py

```diff
@@ -155,215 +371,184 @@ class MyEvent(EventPayload):
-
-        if self._disable:
-            self._logger.debug(
-                "Would have published event",
-                metrics_event=event.model_dump(),
-            )
-            return event
-
+            msg = "Initialize EventManager before creating event publishers"
+            raise EventManagerUnintializedError(msg)
         encoded = await self._schema_manager.serialize(event)
         await publisher.publish(encoded)
-
         self._logger.debug(
             "Published metrics event",
             metrics_event=event.model_dump(),
             topic=publisher.topic,
             schema_info=schema_info,
         )
 
-        return event
+    async def _is_topic_ok(self, topic: str) -> bool:
+        """Check whether the given Kafka topic is valid.
 
-    def _ns_to_datetime(self, ns: int) -> datetime:
-        """Convert a int number of nanoseconds to a datetime."""
-        return datetime.fromtimestamp(ns / 1e9, tz=UTC)
+        Parameters
+        ----------
+        topic
+            Kafka topic to check.
+
+        Returns
+        -------
+        bool
+            `True` if the Kafka topic is valid, `False` otherwise.
+        """
+        info = await self._admin_client.describe_topics([topic])
+        if len(info) == 0:
+            return False
+        topic_info = info[0]
+        error_code = topic_info.get("error_code", None)
+        return error_code == 0
+
+
+class NoopEventManager(EventManager):
+    """An event manager that creates publishers that quietly do nothing.
+
+    This is used as the implementation of `~safir.metrics.EventManager` when
+    event publication is disabled. The event type registrations and event
+    payloads are still verified to catch errors, but are then discarded.
+
+
+    Parameters
+    ----------
+    app_name
+        Name of the application that is generating events.
+    topic_prefix
+        Kafka topic prefix for the metrics events topic for this application.
+    logger
+        Logger to use for internal logging.
+    """
+
+    def __init__(
+        self,
+        app_name: str,
+        topic_prefix: str,
+        logger: BoundLogger | None = None,
+    ) -> None:
+        self._app_name = app_name
+        self._topic = f"{topic_prefix}.{app_name}"
+        self._logger = logger or structlog.get_logger("safir.metrics")
+
+        self._publishers: dict[str, EventPublisher] = {}
+        self._initialized = False
+
+    async def aclose(self) -> None:
+        self._initialized = False
+
+    async def create_publisher(
+        self, name: str, payload_model: type[P]
+    ) -> EventPublisher[P]:
+        if not self._initialized:
+            msg = "Initialize EventManager before creating event publishers"
+            raise EventManagerUnintializedError(msg)
+        if name in self._publishers:
+            raise DuplicateEventError(name)
+
+        # Construct the event model.
+        model = self.build_validated_model(name, self._topic, payload_model)
+
```

Comment:
Is it worth putting this in the base class to avoid duplication and divergence if this logic changes?

Reply:
Yeah, this is a good call, I was being lazy. :) Working on a refactor now.